src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
+        offset: usize,
+        /// The source [`ParseIndexError`]
+        source: ParseIndexError,
+    },
+    /// An `Index` `Token` within the `Pointer` was out of bounds of the
+    /// target array.
+    OutOfBounds {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`OutOfBoundsError`]
+        source: OutOfBoundsError,
+    },
+}
+impl fmt::Display for AssignError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::FailedToParseIndex { offset, .. } => {
+                write!(f, "failed to parse index at offset {}", offset)
+            }
+            Self::OutOfBounds { offset, .. } => {
+                write!(f, "index at offset {} out of bounds", offset)
+            }
+        }
+    }
+}
+#[cfg(feature = "std")]
+impl std::error::Error for AssignError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::FailedToParseIndex { source, .. } => Some(source),
+            Self::OutOfBounds { source, .. } => Some(source),
+        }
+    }
+}
+
+enum Assigned<'v> {
+    Done(Assignment<'v>),
+    Continue {
+        next_buf: PointerBuf,
+        next_value: &'v mut Value,
+        same_src: Value,
+    },
+}
+
+fn assign_array<'v>(
+    token: Token<'_>,
+    remaining: &Pointer,
+    mut buf: PointerBuf,
+    array: &'v mut Vec<Value>,
+    src: Value,
+    offset: usize,
+) -> Result<Assigned<'v>, AssignError> {
+    // parsing the index
+    let idx = token
+        .to_index()
+        .map_err(|source| AssignError::FailedToParseIndex { offset, source })?
+        .for_len_incl(array.len())
+        .map_err(|source| AssignError::OutOfBounds { offset, source })?;
+    buf.push_back(idx.into());
+
+    debug_assert!(idx <= array.len());
+
+    if idx < array.len() {
+        // element exists in the array, we either need to replace it or continue
+        // depending on whether this is the last token or not
+        if remaining.is_root() {
+            // last token, we replace the value and call it a day
+            let replaced = Some(mem::replace(&mut array[idx], src));
+            let assigned = &mut array[idx];
+            Ok(Assigned::Done(Assignment {
+                assigned,
+                assigned_to: buf,
+                replaced,
+            }))
+        } else {
+            // not the last token, we continue with a mut ref to the element as
+            // the next value
+            Ok(Assigned::Continue {
+                next_value: &mut array[idx],
+                same_src: src,
+                next_buf: buf,
+            })
+        }
+    } else {
+        // element does not exist in the array.
+        // we create the path and assign the value
+        let src = expand_src_path(remaining, src)?;
+        array.push(src);
+        // SAFETY: just pushed.
+        let assigned = array.last_mut().unwrap();
```

Comment:
```suggestion
        let assigned = array.last_mut().expect("just pushed");
```
`SAFETY` comments are reserved for unsafe code. Being consistent about this helps tooling check that all unsafe code has been accounted for. Plus using `expect` is generally good practice because it gives context in case the expectation is violated.

Reply:
awesome, thanks